### DISCUSSION TITLE
cmake 中 configure_file 应该如何转化到 xmake表达

### DISCUSSION THREAD

**Discussion by BanKnight** [xmake-io/xmake#2231] [Q&A] [answered]
情况稍稍有点复杂：
要改写的是一个第三方的C语言库[txiki.js](https://github.com/saghul/txiki.js)，
它在cmake中使用configure_file将js复制到build文件夹下
```cmake
configure_file(
        ${CMAKE_SOURCE_DIR}/src/js/bundle.js
        ${CMAKE_CURRENT_BINARY_DIR}/bundle.js
        COPYONLY)
```
然后在编译时，利用编译器特性，嵌入到头文件中，类似于xmake中的utils.bin2c的做法，参考：[incbin](https://github.com/graphitemaster/incbin)
```c
  #include "incbin.h"

  INCTXT(bundle, "bundle.js");
```

我在写xmake测试时，利用rule将js文件复制到build目录下，虽然成功复制过去了，但是编译器还是提示找不到，我观察了一下，
发现提示的目录时/tmp，显然不是build目录。

求助：
这种问题应该如何解决呢？

**Comment by waruqi** [comment, marked as answer]
用 add_configfiles

https://xmake.io/#/zh-cn/manual/project_target?id=targetadd_configfiles

**Reply by BanKnight** [reply]
add_configfiles的作用是“对里面的一些模板变量进行预处理替换”，并不是想要的结果。

需要将js文件复制到正确的编译目录，这样编译器才能正确的将字符串嵌入到头文件里面去。

**Reply by waruqi** [reply]
> add_configfiles的作用是“对里面的一些模板变量进行预处理替换”，并不是想要的结果。
> 
> 需要将js文件复制到正确的编译目录，这样编译器才能正确的将字符串嵌入到头文件里面去。

替换是可选特性，你不设置变量，不就等于直接复制了么，原本涉及初衷，就是支持复制任意文件的。

**Reply by waruqi** [reply]
另外，代码里引用二进制文件，直接用 bin2c rule 就行了，https://xmake.io/#/zh-cn/manual/custom_rule?id=utilsbin2c

**Reply by BanKnight** [reply]
好的，我试下

**Reply by BanKnight** [reply]
发现即使用以下写法，它也还是会去处理
```
add_configfiles("src/js/*.js", {copyonly = true})
```
提示错误：
```
error: cannot get variable(ipv4Part) in src/js/std.js.
```
检查发现，std.js中有这样的内容
```js
      const ipv4Regexes = {
        fourOctet: new RegExp(`^${ipv4Part}\\.${ipv4Part}\\.${ipv4Part}\\.${ipv4Part}$`, "i"),
        threeOctet: new RegExp(`^${ipv4Part}\\.${ipv4Part}\\.${ipv4Part}$`, "i"),
        twoOctet: new RegExp(`^${ipv4Part}\\.${ipv4Part}$`, "i"),
        longValue: new RegExp(`^${ipv4Part}$`, "i")
      };
```

btw，之所以不用bin2c rule，是想要尽量不改动他的代码

**Reply by waruqi** [reply]
文档里写错了，应该是 onlycopy
